File: backend/app/schemas/document.py
"""
Document Schemas

Pydantic models for document API request/response validation
"""

from pydantic import BaseModel, Field, ConfigDict, field_serializer
from datetime import datetime
from typing import Optional

from app.models.document import ProcessingStatus


# Request Schemas

class DocumentUploadResponse(BaseModel):
    """Response after successful document upload"""
    id: int
    filename: str
    file_type: str
    file_size: int
    status: ProcessingStatus
    upload_date: datetime
    message: str = "Document uploaded successfully"
    
    @field_serializer('status')
    def serialize_status(self, status: ProcessingStatus) -> str:
        """Convert status enum to lowercase for frontend consistency"""
        return status.value.lower()
    
    model_config = ConfigDict(from_attributes=True)


# Response Schemas

class DocumentResponse(BaseModel):
    """Complete document information"""
    id: int
    user_id: int
    filename: str
    file_path: str
    file_type: str
    file_size: int
    upload_date: datetime
    last_modified: datetime
    status: ProcessingStatus
    error_message: Optional[str] = None
    page_count: Optional[int] = None
    word_count: Optional[int] = None
    
    # Computed properties
    file_size_mb: float
    is_ready_for_query: bool
    
    @field_serializer('status')
    def serialize_status(self, status: ProcessingStatus) -> str:
        """Convert status enum to lowercase for frontend consistency"""
        return status.value.lower()
    
    model_config = ConfigDict(from_attributes=True)


class DocumentListResponse(BaseModel):
    """Paginated list of documents"""
    documents: list[DocumentResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class DocumentSummary(BaseModel):
    """Brief document information for list views"""
    id: int
    filename: str
    file_type: str
    file_size_mb: float
    upload_date: datetime
    status: ProcessingStatus
    error_message: Optional[str] = None
    is_ready_for_query: bool
    store_name: Optional[str] = None
    
    @field_serializer('status')
    def serialize_status(self, status: ProcessingStatus) -> str:
        """Convert status enum to lowercase for frontend consistency"""
        return status.value.lower()
    
    model_config = ConfigDict(from_attributes=True)


class DocumentListSummary(BaseModel):
    """Paginated list of document summaries"""
    documents: list[DocumentSummary]
    total: int
    page: int
    page_size: int
    total_pages: int


class DocumentDeleteResponse(BaseModel):
    """Response after document deletion"""
    message: str
    document_id: int
    filename: str


class DocumentStats(BaseModel):
    """User's document statistics"""
    total_documents: int
    total_size_mb: float
    by_status: dict[str, int]
    by_type: dict[str, int]

File: frontend/src/types/document.ts
/**
 * Document Types
 * 
 * TypeScript interfaces matching backend Pydantic schemas
 */

export type ProcessingStatus = "uploaded" | "processing" | "completed" | "failed";

export const ProcessingStatus = {
  UPLOADED: "uploaded" as const,
  PROCESSING: "processing" as const,
  COMPLETED: "completed" as const,
  FAILED: "failed" as const,
};

export interface Document {
  id: number;
  filename: string;
  file_type: string;
  file_size?: number;
  file_size_mb: number;
  s3_path?: string;
  status: ProcessingStatus;  // Changed from processing_status to match backend
  upload_date: string;
  processed_at?: string;
  error_message?: string;
  user_id?: number;
  // Computed properties
  is_ready_for_query: boolean;
}

export interface DocumentUploadResponse {
  id: number;
  filename: string;
  original_filename: string;
  file_type: string;
  file_size: number;
  file_size_mb: number;
  processing_status: ProcessingStatus;
  upload_date: string;  // Changed from uploaded_at to upload_date
  message: string;
}

export interface DocumentListResponse {
  documents: Document[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface DocumentStats {
  total_documents: number;
  total_size_bytes: number;
  total_size_mb: number;
  by_status: {
    uploaded: number;
    processing: number;
    completed: number;
    failed: number;
  };
  by_type: Record<string, number>;
}

export interface UploadProgress {
  loaded: number;
  total: number;
  percentage: number;
}
